Rework user integrity checks for Auditus

hanneskuettner opened the pull request:
<!--

Heya! Thanks for opening a Pull Request! If your PR is implementing a new feature or fix for Directus, please make sure your PR adheres to the following requirements:

- The PR closes an Issue (not Discussion)
- Tests are added/updated and are passing locally if applicable
- Documentation was added/updated if applicable

Please make sure to "Link" the issue you're closing. Without a Linked issue, this PR won't be accepted. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.

-->

## Scope

The general idea behind this change is to delay the user integrity checks (user limits & remaining admin users) to after all changes have been made to the DB within a transaction and abort that transaction if the integrity is not given. But since modification to system collection that are relevant for the user count (per access level) can happen at any point of a mutation and it only makes sense to do the check once all mutations have been applied, it is necessary to perform this check in the `ItemsService` itself.

For this to work, every individual item service needs a way of marking a mutation as relevant for triggering a user integrity check based on the contents of the revision. For example, the role service needs to trigger and integrity check if the `parent` of a role changed, since that can influence the access types of all users.

Additionally it is useful to only do certain checks for some mutations, for example creating new users does not require us to validate the number of existing admin users > 0, neither does the granting of additional app or admin rights in a policy. So we want some control over what integrity checks need to be performed.

This change assumes that all top level system services are called correctly from nested payload processing operations, because otherwise we don't have a chance of identifying nested changes. This will be ensured in a different PR from @paescuj.

**What's changed:**

**`ItemsService`**
- Introduce two new options to `MutationOptions`
  - `userIntegrityCheckFlags` indicate if, and what user integrity checks need to be performed
  - `onRequireUserIntegrityCheck` a callback a nested service can call in order to pass up additional user integrity check flags
- Trigger a user integrity count in the top level `ItemsService` mutation (identified by not having a `onRequireUserIntegrityCheck` passed in) if any nested service identifies the need for a check
- Additionally system services that inherit from `ItemsService` can pass in the `userIntegrityCheckFlags` if a check is deemed necessary from a top level call

**User counting**
- This change vastly simplifies the user limit checking, in that way that we don't need to pre-calculate expected changes in the users and roles service, but instead to a targeted recount if necessary
- This allowed me to remove a lot of the `get-*-count` helpers
- Reworked `fetchUserCount` to accept roles, users, policies and access rows to ignore, in order to allow for a reduced admin check if some of those have been removed
- Reworked `fetchUserCount` to use aggregation instead of iterating over all users sequentially

**System Services**
- Selectively trigger a user integrity check if something related to user counts or access has changed, to avoid, for example having to do a recount on every `last_page` user update

## Potential Risks / Drawbacks

- Some, probably

## Review Notes / Questions

- Does this architecture make sense?
- Are we missing updating the flags somewhere?
- Are we too selective of when to do a recount?
- Some telemetry tests are still failing (this I will fix) and user and roles service tests are even more borked now (this I won't (yet))

hanneskuettner committed:
Changes to user counting and integrity checks

hanneskuettner committed:
Ensure that user validation happens in both create one and create many

hanneskuettner committed:
Rename `checkType` to `flags`

DanielBiegler reviewed (commented):
Partial review but includes practical suggestions for the bit flags

hanneskuettner commented on api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts:
```diff
@@ -0,0 +1,9 @@
+import { UnprocessableContentError } from '@directus/errors';
+
+export function validateRemainingAdminCount(count: number) {
```
We use `validate` a lot more often now in the auditus rework, e.g. `validatePath`, `validateAccess`, etc. and all of those are throwing errors if something is wrong. So I think sticking with the `validate` is fine :)

hanneskuettner committed:
Update api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

hanneskuettner commented on api/src/utils/validate-user-count-integrity.ts:
```diff
@@ -0,0 +1,41 @@
+import { shouldCheckUserLimits } from '../telemetry/utils/should-check-user-limits.js';
+import { fetchUserCount, type FetchUserCountOptions } from './fetch-user-count/fetch-user-count.js';
+
+enum UserIntegrityCheckType {
+	/** Check if the number of remaining admin users is greater than 0 */
+	RemainingAdmins = 0,
+	/** Check if the number of users is within the limits */
+	UserLimits = 1,
+}
+
+export const USER_INTEGRITY_CHECK_REMAINING_ADMINS = 1 << UserIntegrityCheckType.RemainingAdmins;
+export const USER_INTEGRITY_CHECK_USER_LIMITS = 1 << UserIntegrityCheckType.UserLimits;
+export const USER_INTEGRITY_CHECK_ALL = USER_INTEGRITY_CHECK_REMAINING_ADMINS | USER_INTEGRITY_CHECK_USER_LIMITS;
+
+export interface ValidateUserCountIntegrityOptions extends Omit<FetchUserCountOptions, 'adminOnly'> {
+	flags: number;
+}
```
Oh that enum suggestion is super sweet and totally makes sense, didn't even cross my mind, I guess that is polluted with old C stuff ha.

I think sticking to the flags, with your suggested changes makes the most sense, since that makes it easy to just "pile on" to existing flags using the bit-wise or. Otherwise we would need to always check against the already set flag, when updating the flags.

I currently don't see a reason for splitting it up into more granular checks, as a user limit check always needs to include all users and the remaining admin check is a functionally different check from something that would fall under an `Admin` flag. Does that make sense 🤔?

I'm keeping this as unresolved for now, for others to chime in.

paescuj replied:
Great work ❤️ Makes sense as it is now for me 👌

hanneskuettner committed:
Update to enum usage

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

hanneskuettner committed:
A few more changes to enum instead of number

hanneskuettner committed:
One more enum type update

hanneskuettner committed:
Make sure to correctly override the callback when combining options

hanneskuettner committed:
Clean up option type

licitdev commented on api/src/services/users.ts:
```diff
@@ -161,18 +158,41 @@ export class UsersService extends ItemsService {
+		try {
+			if ('email' in data) {
+				this.validateEmail(data['email']);
+				await this.checkUniqueEmails(data['email']);
```
An error occurs during user creation as the email passed in isn't an array.

We could either wrap with an array as follows or use `toArray()` within `checkUniqueEmails()`

```suggestion
				await this.checkUniqueEmails([data['email']]);
```

hanneskuettner committed:
Update api/src/services/users.ts

Co-authored-by: ian <licitdev@gmail.com>

hanneskuettner committed:
Only take validation shortcut for users

We can only be sure that the deletion of users does not increase any other access types count, so in all other cases we need to verify that for example the App or API users have not increased over the limit

hanneskuettner committed:
Make both app and admin users count against app access limit

paescuj commented on api/src/services/access.ts:
```diff
@@ -22,18 +28,30 @@ export class AccessService extends ItemsService {
 	override async deleteMany(keys: PrimaryKey[], opts: MutationOptions = {}): Promise<PrimaryKey[]> {
+		// Changes here can affect the number of admin/app/api users
+		if (opts.onRequireUserIntegrityCheck) {
+			opts.onRequireUserIntegrityCheck(UserIntegrityCheckFlag.All);
```
Do we really need to check `All` here? Wouldn't remaining admins be sufficient?

```suggestion
			opts.onRequireUserIntegrityCheck(UserIntegrityCheckFlag.RemainingAdmins);
```

hanneskuettner replied:
Sadly not, I thought so too at first, but deleting a policy can move users from app or admin users to api users, which have a separate limit which might then be breached.

paescuj replied:
I seeee ✔️

paescuj commented on api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts:
```diff
@@ -0,0 +1,9 @@
+export function validateRemainingAdminCount(count: number) {
+	if (count <= 0) {
+		throw new UnprocessableContentError({
+			reason: `Doing this would remove the last admin user from the system`,
```
"Doing this" sounds a bit too generic IMO 😇 
```suggestion
			reason: `Cannot remove the last admin user from the system`,
```

paescuj commented on api/src/services/policies.ts:
```diff
@@ -33,6 +84,12 @@ export class PoliciesService extends ItemsService<Policy> {
+		if (opts.onRequireUserIntegrityCheck) {
+			opts.onRequireUserIntegrityCheck(UserIntegrityCheckFlag.All);
+		} else {
+			opts.userIntegrityCheckFlags = UserIntegrityCheckFlag.All;
```
Same question as in https://github.com/directus/directus/pull/22737/files#r1639939752

hanneskuettner replied:
Same answer. 😆

hanneskuettner commented on api/src/services/items.ts:
```diff
@@ -706,18 +752,33 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
 
 				childrenRevisions.push(...revisions);
 				nestedActionEvents.push(...nestedActionEventsO2M);
+				userIntegrityCheckFlags ||= userIntegrityCheckFlagsO2M;
```
Fix this @hanneskuettner! Now! Needs to be `|=`

paescuj replied:
![gogo](https://github.com/directus/directus/assets/5363448/91a083f1-1703-4ce5-8655-947ccfff35ca)

paescuj commented on api/src/telemetry/utils/check-user-limits.ts:
I know, these utils had been located in `telemetry` before, but this doesn't seem the correct place. Could we move that? Probably to the global `utils` folder where the validation function lives as well.

hanneskuettner replied:
Agreed. Also doesn't seem like the right place. I'll probably also move `validate-remaining-admin-*` in there.

paescuj replied:
Awesome!

hanneskuettner commented on api/src/services/users.ts:
```diff
@@ -161,18 +158,41 @@ export class UsersService extends ItemsService {
+			}
+
+			if ('password' in data) {
+				await this.checkPasswordPolicy(data['password']);
```
This also needs to be `[data['password']]`. Fix this @hanneskuettner

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-rework-user-counting

hanneskuettner committed:
Update api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

hanneskuettner committed:
One post-merge fix, two small fixes

hanneskuettner committed:
Simplify flag updating and callback calling

hanneskuettner committed:
Changing app access in a policy only requires user limit checking, not full check

hanneskuettner committed:
Only the status of a created user should matter to determine if a check is neccessary

rijkvanzanten commented on api/src/utils/fetch-user-count/get-user-count-query.ts:
```diff
@@ -0,0 +1,31 @@
+		return Promise.resolve({ count: 0 });
+	}
+
+	let query = knex('directus_users').count({ count: '*' }).where('status', 'active');
```
I believe this needs an `.as('count')`. Some database vendors will auto-name the column `count(*)` instead of `count`

hanneskuettner replied:
A bit late, but isn't that the alias syntax for knex if you throw in a object, where the keys are used as aliases?

rijkvanzanten replied:
TIL! Didn't know knex did it that way (their api can be a bit.. interesting at times)

rijkvanzanten commented on api/src/utils/validate-user-count-integrity.ts:
```diff
@@ -0,0 +1,39 @@
+export enum UserIntegrityCheckFlag {
+	None = 0,
+	/** Check if the number of remaining admin users is greater than 0 */
+	RemainingAdmins = 1 << 0,
+	/** Check if the number of users is within the limits */
+	UserLimits = 1 << 1,
+	All = ~(~0 << 2),
```
This is pure black magic

rijkvanzanten commented on api/src/utils/fetch-user-count/fetch-access-lookup.ts:
```diff
@@ -17,4 +27,30 @@ export async function fetchAccessLookup(knex: Knex): Promise<AccessLookup[]> {
+	}
+
+	if (options.adminOnly) {
+		query = query.where('directus_policies.admin_access', 1);
```
Seems good in Postgres; we'll do more in-depth testing once this is in the v11-rc state

hanneskuettner replied:
Good call!

rijkvanzanten committed:
Add count alias to count query

rijkvanzanten reviewed (approved):
LGTM
